pkg/cache/clusterqueue.go

```diff
@@ -120,12 +125,11 @@ func newCohort(name string, size int) *Cohort {
 	}
 }
 
-func (c *Cohort) CanFit(q FlavorResourceQuantities) bool {
+func (c *ClusterQueue) FitInCohort(q FlavorResourceQuantities) bool {
 	for flavor, qResources := range q {
-		if cohortResources, flavorFound := c.RequestableResources[flavor]; flavorFound {
-			cohortUsage := c.Usage[flavor]
+		if _, flavorFound := c.Cohort.RequestableResources[flavor]; flavorFound {
 			for resource, value := range qResources {
-				available := cohortResources[resource] - cohortUsage[resource]
+				available := c.RequestableCohortQuota(flavor, resource) - c.UsedCohortQuota(flavor, resource)
```

Comment:
I would prefer we call the feature gate here, preserving the old code.

And these functions really work together, so it might be better to have a single function called `c.AvailableQuotaWithCohortBorrowing(flavor, resource)`

Reply:
Also, an alternative implementation is as follows:

1. `cohort.RequestableResources` only contains the sum of lendingLimits (same as your approach).
2. `cohort.Usage` only contains the usage within the `RequestableResources` as defined in 1 (same as your approach).
3. When evaluating a workload admission, we calculate: `excess := workload.request-guaranteed`. The workload fits if `cohort.Usage+excess <= cohort.RequestableResources`.

Both implementations are equivalent, but I think this is easier to document and follow along.

Reply:
> And these functions really work together, so it might be better to have a single function called `c.AvailableQuotaWithCohortBorrowing(flavor, resource)`

`RequestableCohortQuota` could not be moved into a single function: [here](https://github.com/kubernetes-sigs/kueue/pull/1385/files#diff-0870ec2c0ea330a7d087a6ef599905bf9b7164a0dd889d1b00ba17c31a1bfd05L563) `cohortAvailable` needs to be calculated individually. And `UsedCohortQuota` [here](https://github.com/kubernetes-sigs/kueue/pull/1385/files#diff-0870ec2c0ea330a7d087a6ef599905bf9b7164a0dd889d1b00ba17c31a1bfd05R575).

Reply:
> I would prefer we call the feature gate here, preserving the old code.

We can change this part like this:
```
for flavor, qResources := range q {
	if cohortResources, flavorFound := c.RequestableResources[flavor]; flavorFound {
		cohortUsage := c.Usage[flavor]
		for resource, value := range qResources {
			available := cohortResources[resource] - cohortUsage[resource]
			if features.Enabled(features.LendingLimit) {
				available = c.RequestableCohortQuota(flavor, resource) - c.UsedCohortQuota(flavor, resource)
			}
```
But since we already have featureGates inside of `c.RequestableCohortQuota` and `c.UsedCohortQuota`([RequestableCohortQuota](https://github.com/kubernetes-sigs/kueue/pull/1385/files#diff-441da8393964615cf5b020d0afb9199588e157e50718e83b527788eb679391ecR537), [guaranteedQuota](https://github.com/kubernetes-sigs/kueue/pull/1385/files#diff-441da8393964615cf5b020d0afb9199588e157e50718e83b527788eb679391ecR543), [UsedCohortQuota](https://github.com/kubernetes-sigs/kueue/pull/1385/files#diff-441da8393964615cf5b020d0afb9199588e157e50718e83b527788eb679391ecR565)), is it still necessary to add featureGates here?

Reply:
> But since we already have featureGates inside of c.RequestableCohortQuota and c.UsedCohortQuota(RequestableCohortQuota, guaranteedQuota, UsedCohortQuota), is it still necessary to add featureGates here?

The fact that the usage of the feature gate is scattered among multiple functions makes it hard to visually verify if the old behavior is preserved. It's generally a good practice to use feature gates in a coarse manner.

You could actually remove the feature gate checks from the new functions and check the feature gate in the flavorassigner code as well.